whitepaper/update.py

```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/env python3
+from pathlib import Path
+import shutil
+import base64
+import subprocess
+from subprocess import run
+from typing import List
+import re
+import textwrap
+import tempfile
+from typing import Optional
+import functools
+import asyncio
+
+import typer
+import toml
+import pydantic
+import rich.console
+import rich.panel
+import rich.pretty
+from jinja2 import Template
+import aiohttp
+from gidgethub.aiohttp import GitHubAPI
+from fsspec.implementations.zip import ZipFileSystem
+
+
+app = typer.Typer()
+
+
+class MetaData(pydantic.BaseModel):
+    authors: List[str]
+    title: str
+    description: str = ""
+
+
+class WhitePaper(pydantic.BaseModel):
+    repository: str
+    metadata: MetaData | None = None
+
+    @property
+    def repo(self) -> str:
+        m = re.match(r"^https://github.com/(.*/.*)$", self.repository)
+        assert m is not None
+        return m.group(1)
+
+    @property
+    def slug(self) -> str:
+        owner, name = self.repo.split("/")
+        return f"{owner}_{name}"
+
+
+class Config(pydantic.BaseModel):
+    whitepapers: List[WhitePaper]
+
+
+def parse_metadata(content: str) -> MetaData:
+    m = re.search(r"\\author{(.*)}", content)
+    assert m is not None
+    author_text = m.group(1)
+    authors = [a.strip() for a in author_text.split(r"\and")]
+
+    m = re.search(r"\\title{(.*)}", content)
+    assert m is not None
+    title = m.group(1).strip()
+
+    return MetaData(authors=authors, title=title)
+
+
+def which(cmd: str) -> Optional[Path]:
+    try:
+        exe = (
+            run(["command", "-v", cmd], check=True, capture_output=True)
+            .stdout.decode()
+            .strip()
+        )
+        return Path(exe)
+    except subprocess.CalledProcessError:
+        return None
+
+
+def coro(fn):
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        return asyncio.run(fn(*args, **kwargs))
+
+    return wrapper
+
+
+def make_titlepage_image(convert: Path, pdf: Path, output: Path):
+    run(
+        [
+            convert,
+            "-density",
+            "300",
+            f"{pdf}[0]",
+            "-background",
+            "white",
+            "-alpha",
+            "remove",
+            "-alpha",
+            "off",
+            "-depth",
+            "2",
+            str(output),
+        ],
+        check=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+    )
+
+
+async def get_file(gh: GitHubAPI, repo: str, file: str) -> str:
+    r = await gh.getitem(f"/repos/{repo}/contents/{file}")
+    return base64.b64decode(r["content"]).decode()
+
+
+@app.command()
+@coro
+async def pull(
```

Comment:
Depends on whether it's preserved when it's manually added. I don't think there's much we can do since the file content is generated by the toml library, and as far as I know it's not possible to change the output.